specs/bep-v1.rst

```diff
@@ -62,468 +62,274 @@ another.
-request messages may be outstanding at any given moment. For message types
-that do not have a corresponding response (Cluster Configuration, Index, etc.)
-the Message ID field is irrelevant and SHOULD be set to zero.
-
-The **Type** field indicates the type of data following the message header
-and is one of the integers defined below. A message of an unknown type
-is a protocol error and MUST result in the connection being terminated.
-
-The **Compression** bit "C" indicates the compression used for the message.
-
-For C=0:
-
--  The Length field contains the length, in bytes, of the uncompressed
-   message data.
-
--  The message is not compressed.
-
-For C=1:
-
--  The Length field contains the length, in bytes, of the compressed
-   message data plus a four byte uncompressed length field.
-
--  The compressed message data is preceded by a 32 bit field denoting
-   the length of the uncompressed message.
-
--  The message data is compressed using the LZ4 format and algorithm
-   described in http://www.lz4.org/.
-
-All data within the message (post decompression, if compression is in
-use) MUST be in XDR (RFC 1014) encoding. All fields shorter than 32 bits
-and all variable length data MUST be padded to a multiple of 32 bits.
-The actual data types in use by BEP, in XDR naming convention, are the
-following:
+Every message post authentication is made up of several parts:
+
+- A header length word
+- A **Header**
+- A message length word
+- A **Message**
+
+.. code-block:: none
+
+     0                   1
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |         Header Length         |
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /                               /
+    \            Header             \
+    /                               /
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |         Message Length        |
+    |           (32 bits)           |
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /                               /
+    \            Message            \
+    /                               /
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+The header length word is 16 bits. It indicates the length of the following
+**Header** message. The Header is in protocol buffer format. The Header
+describes the type and compression status of the following message.
+
+The message is preceded by the 32 bit message length word and is one of the
+concrete BEP messages described below, identified by the **type** field of
+the Header.
+
+.. code-block:: proto
+
+    message Header {
+        MessageType        type        = 1;
+        MessageCompression compression = 2;
+    }
 
-(unsigned) int:
-    (unsigned) 32 bit integer
+    enum MessageType {
+        CLUSTER_CONFIG    = 0;
+        INDEX             = 1;
+        INDEX_UPDATE      = 2;
+        REQUEST           = 3;
+        RESPONSE          = 4;
+        DOWNLOAD_PROGRESS = 5;
+        PING              = 6;
+        CLOSE             = 7;
+    }
```

Comment:
I highly recommend that enum types have sane default values. In most cases, I think that should be `UNKNOWN = 0`.

If a client doesn't understand a message, they'll be able to tell they can't understand it, instead of not being able to parse the message.

https://developers.google.com/protocol-buffers/docs/proto3#default